### ISSUE TITLE
Can't statically link curl on mingw64

### ISSUE THREAD

**Issue by sunbinyuan** [curl/curl#5023]
### I did this
> CFLAGS=-DCURL_STATICLIB gcc test.c -lcurl

where `test.c` is the example *Download many files in parallel, in the same thread.*
```
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x103): undefined reference to `__imp_curl_easy_init'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x128): undefined reference to `__imp_curl_easy_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x15d): undefined reference to `__imp_curl_easy_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x192): undefined reference to `__imp_curl_easy_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x1a6): undefined reference to `__imp_curl_multi_add_handle'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x1e5): undefined reference to `__imp_curl_global_init'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x1ee): undefined reference to `__imp_curl_multi_init'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x20d): undefined reference to `__imp_curl_multi_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x243): undefined reference to `__imp_curl_multi_perform'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x288): undefined reference to `__imp_curl_easy_getinfo'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x29e): undefined reference to `__imp_curl_easy_strerror'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x2e4): undefined reference to `__imp_curl_multi_remove_handle'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x2f4): undefined reference to `__imp_curl_easy_cleanup'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x34d): undefined reference to `__imp_curl_multi_info_read'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x38d): undefined reference to `__imp_curl_multi_wait'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x3b2): undefined reference to `__imp_curl_multi_cleanup'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccJlgl3R.o:test.c:(.text+0x3bb): undefined reference to `__imp_curl_global_cleanup'
collect2.exe: error: ld returned 1 exit status
```
I also tried to manually link the prebuilt Windows x64 binaries from [here](https://curl.haxx.se/windows/) by extracting `lib` and `include` on the root folder and running `CFLAGS=-DCURL_STATICLIB gcc -static test.c -Iinclude -Llib -lcurl`. 

**Also note that I was able to built both of them successfully once I remove the `-static` flag.**
### I expected the following
Application to compile with curl statically linked.
### curl/libcurl version

```
curl 7.68.0 (x86_64-w64-mingw32) libcurl/7.68.0 OpenSSL/1.1.1d (Schannel) zlib/1.2.11 brotli/1.0.7 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.40.0Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smtp smtps telnet tftp
Features: AsynchDNS brotli HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz Metalink MultiSSL NTLM PSL SPNEGO SSL SSPI TLS-SRP
```
compiled with the `PKGCONFIG`  with flags `--winssl` from the [MSYS2 repo](https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-curl) .
### operating system
Windows 10 (64bit)

**Comment by jay** [comment]
> CFLAGS=-DCURL_STATICLIB gcc test.c -lcurl

hm I don't know that gcc uses CFLAGS... also does msys2 give you a curl-config? It is a lot easier to build static if you use that.

try this

gcc test.c -DCURL_STATICLIB `curl-config --cflags --static-libs`

**Comment by sunbinyuan** [comment]
It worked like a charm, I am completely baffled. I looked everywhere about this issue but could not find any resources about it. I looked at the `INSTALL.md` page but the only thing they mention about static linking in Windows was to include `-DCURL_STATICLIB`. Maybe having it in there might help others with the same issue as me. Thank you.

**Comment by bagder** [comment]
INSTALL.md is however a description and documentation for how to build and install curl and libcurl, it is not a tutorial for how you build your applications with libcurl.

The [libcurl-tutorial.3 man page](https://curl.haxx.se/libcurl/c/libcurl-tutorial.html) should probably be a more suitable place, and it already documents and references `curl-config`. If you can think of improvements there, feel free to provide a PR!

**Comment by sunbinyuan** [comment]
This might be a seperate issue but I just realized that running ``gcc test.c -DCURL_STATICLIB `curl-config --cflags --static-libs` ``, althought compiled, did not made it static yet. I then ran it with ``gcc test.c -static -DCURL_STATICLIB `curl-config --cflags --static-libs` `` but I am now having an issue with the linker having a different issue `C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lbrotlidec`. I am not sure if this would still an issue with building curl or rather with the brotlidec library.

**Comment by bagder** [comment]
I don't see how it curl's fault if the linker cannot find the brotli decompression library (presumably a static version of it).

**Comment by jay** [comment]
> ld.exe: cannot find -lbrotlidec

It's a bug in brotli (in my opinion), see google/brotli#795

**Comment by rajhlinux** [comment]
I'm getting this error, similar to OP:

C:\Users\Admin\Downloads\Compressed\web_dev>gcc simple.c -L"C:\Users\Admin\Downloads\Compressed\web_dev\aaaaaaaaa\lib" -lcurl
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x128): undefined reference to `__imp_curl_easy_init'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x158): undefined reference to `__imp_curl_easy_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x178): undefined reference to `__imp_curl_easy_setopt'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x188): undefined reference to `__imp_curl_easy_perform'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x19f): undefined reference to `__imp_curl_easy_strerror'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\Admin\AppData\Local\Temp\ccrlPnr7.o:simple.c:(.text+0x1d2): undefined reference to `__imp_curl_easy_cleanup'
collect2.exe: error: ld returned 1 exit status

I tried this and it doesn't work:
gcc simple.c -DCURL_STATICLIB `curl-config --cflags --static-libs`

**Comment by jay** [comment]
> I tried this and it doesn't work:
> gcc simple.c -DCURL_STATICLIB \`curl-config --cflags --static-libs\`

Ok but _how_ doesn't it work? Please open a new issue and include the output of gcc and `curl-config --cflags --static-libs`
